src/hotspot/share/opto/cfgnode.hpp

```diff
@@ -496,17 +501,11 @@ class ParsePredicateNode : public IfNode {
     return _deopt_reason;
   }
 
-  bool is_useless() const {
-    return _useless;
-  }
-
-  void mark_useless() {
-    _useless = true;
-  }
-
-  void mark_useful() {
-    _useless = false;
-  }
+  bool is_useless() const;
+  void mark_useless(PhaseIterGVN& igvn);
+  void mark_maybe_useful();
+  bool is_useful() const;
+  void mark_useful();
```

Comment:
Needed to move these definitions to the source file because I cannot include `predicates.hpp` here due to circular dependencies. I solved this by forward declaring `PredicateState` and moving the definitions to the source file.

Same for these methods for `OpaqueTemplateAssertionPredicate` further down.

Reply:
Can you add `predicates.inline.hpp` for this?

Reply:
Do you mean the enum definition for `PredicateState`? That could work, so I can just include that header when I need to use the enum. Should I then rather name the `hpp`  file something like `predicates_enums.hpp`? IIUC, `xyz.inline.hpp` should be used primarily for inline methods.

Having the separate enum header also allows us to put this one there which makes things cleaner:
https://github.com/openjdk/jdk/blob/e3c29c9e6cff7648952c0ba359b0763a0ea8da18/src/hotspot/share/opto/predicates.hpp#L203-L211

Reply:
I've pushed an update introducing `predicates_enums.hpp` and also moving some typedefs for predicates in there. Let me know if that's what you had in mind. I could now move almost all definitions back to the header file except for `mark_useless()` which uses `PhaseIterGVN` which is not a complete type in the header file - I guess that's okay to just move this one to the source file.